splight_lib/models/datalake_base.py

```diff
@@ -0,0 +1,138 @@
+    ) -> list[Self]:
+        request = _to_data_request(
+            cls, asset, attribute, extra_pipeline, **params
+        )
+        request.add_trace(Trace.from_address(asset, attribute))
+        return request.apply()
+
+    @classmethod
+    async def async_get(
+        cls,
+        asset: str | Asset,
+        attribute: str | Attribute,
+        extra_pipeline: list[dict[str, Any]] = {},
+        **params: Dict,
+    ) -> list[Self]:
+        request = _to_data_request(
+            cls, asset, attribute, extra_pipeline, **params
+        )
+        request.add_trace(Trace.from_address(asset, attribute))
+        instances = await request.async_apply()
+        return instances
+
+    @classmethod
+    def get_dataframe(
+        cls,
+        asset: str | Asset,
+        attribute: str | Attribute,
+        extra_pipeline: list[dict[str, Any]] = {},
+        **params: Dict,
+    ) -> pd.DataFrame:
+        request = _to_data_request(
+            cls, asset, attribute, extra_pipeline, **params
+        )
+        instances = request.apply()
+        df = pd.DataFrame([instance.dict() for instance in instances])
+        return df
+
+    def save(self) -> None:
+        records = self._to_record()
+        records.apply()
+
+    async def async_save(self) -> None:
+        records = self._to_record()
+        await records.async_apply()
+
+    @classmethod
+    def save_dataframe(cls, df: pd.DataFrame):
+        df = _fix_dataframe_timestamp(df)
+        instances = df.to_dict("records")
+        records = DataRecords(
+            collection=cls._collection_name,
+            records=instances,
+        )
+        records.apply()
+
+    def dict(self, *args, **kwargs) -> Dict:
+        d = super().model_dump(*args, **kwargs)
+        return {
+            k: v["id"] if isinstance(v, dict) and "id" in v.keys() else v
+            for k, v in d.items()
+        }
+
+    def _to_record(self) -> DataRecords:
+        return DataRecords(
+            collection=self._collection_name,
+            records=[self.model_dump(mode="json")],
+        )
+
+
+def _to_data_request(
+    model_class: TypeVar("T"),
+    asset: str | Asset,
+    attribute: str | Attribute,
+    extra_pipeline: list[dict[str, Any]] = {},
+    **params: Dict,
+) -> DataRequest:
+    if not isinstance(extra_pipeline, list):
+        raise ValueError("extra_pipeline must be a list of dicts")
+    request = DataRequest[model_class](
+        from_timestamp=params.get("from_timestamp"),
+        to_timestamp=params.get("to_timestamp"),
+    )
+    trace = Trace.from_address(asset, attribute)
+    for step in extra_pipeline:
+        trace.add_step(PipelineStep.from_dict(step))
+    request.add_trace(trace)
+    return request
+
+
+def datalake_model_serializer(data: Dict, default=str, **dumps_kwargs):
+    new_data = {
+        k: v if not isinstance(v, dict) else v["id"] for k, v in data.items()
+    }
+    return json.dumps(new_data, default=default, **dumps_kwargs)
```

Comment:
Where is this used?

Reply:
Is not being used anymore, I can remove it